### PR TITLE
klee: make llvmPackages and uclibc overridable

### PR DESCRIPTION
**Pull request by numinit**
## Description of changes

This should allow users to pick which llvmPackages and uclibc they would like to use, since Klee supports llvmPackages_11 through llvmPackages_16 as of Klee 3.1 (defaulting to 13 for stability).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/science/logic/klee/default.nix [hunk 1]
```diff
@@ -37,12 +35,8 @@
...
   # Python used for KLEE tests.
   kleePython = python3.withPackages (ps: with ps; [ tabulate ]);
-
-  # The klee-uclibc derivation.
-  kleeuClibc = callPackage ./klee-uclibc.nix {
-    inherit stdenv clang llvm extraKleeuClibcConfig debugRuntime runtimeAsserts;
-  };
-in stdenv.mkDerivation rec {
+in
+llvmPackages.stdenv.mkDerivation (final: rec {
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:39]: Why `final: rec {` instead of just `rec`?

File: pkgs/applications/science/logic/klee/default.nix [hunk 2]
```diff
@@ -104,7 +100,9 @@ in stdenv.mkDerivation rec {
 
   passthru = {
     # Let the user depend on `klee.uclibc` for klee-uclibc
-    uclibc = kleeuClibc;
+    uclibc = callPackage ./klee-uclibc.nix {
+      inherit llvmPackages extraKleeuClibcConfig debugRuntime runtimeAsserts;
+    };
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:105]: Could make this an attribute which can be passed. Could even give it a default.

File: pkgs/applications/science/logic/klee/default.nix [hunk 3]
```diff
@@ -128,9 +126,9 @@ in stdenv.mkDerivation rec {
       that matches a computed test input, including setting up files, pipes,
       environment variables, and passing command line arguments.
     '';
-    homepage = "https://klee.github.io/";
+    homepage = "https://github.com/klee/klee";
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:129]: Why the change from the website to the GitHub for the homepage?

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -32396,11 +32396,9 @@ with pkgs;
 
   klayout = libsForQt5.callPackage ../applications/misc/klayout { };
 
-  klee = callPackage ../applications/science/logic/klee (with llvmPackages_13; {
-    clang = clang;
-    llvm = llvm;
-    stdenv = stdenv;
-  });
+  klee = callPackage ../applications/science/logic/klee {
+    llvmPackages = llvmPackages_13;
```
Comment (RossComputerGuy) [inline pkgs/top-level/all-packages.nix:32400]: I see Klee does support LLVM 16 but uses 13 as the default. Why don't we use 16?

File: pkgs/applications/science/logic/klee/default.nix (hunk 3 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix:129]: Consistency with klee-uclibc

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (numinit) [reply pkgs/top-level/all-packages.nix:32400]: 13 is recommended: https://github.com/klee/klee/releases/tag/v3.1

File: pkgs/applications/science/logic/klee/default.nix (hunk 1 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix:39]: It was to facilitate the passthru but I can remove this.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (RossComputerGuy) [reply pkgs/top-level/all-packages.nix:32400]: Alright. Just add an assert check in Klee to make sure that someone can't just pass `llvmPackages_12`.

File: pkgs/applications/science/logic/klee/default.nix (hunk 3 above)
Comment (RossComputerGuy) [reply pkgs/applications/science/logic/klee/default.nix:129]: Oh, though would most people need to know the GitHub or would they want to visit the website for finding info on klee?

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (numinit) [reply pkgs/top-level/all-packages.nix]: Min is actually 11, so made that work.

**Comment by numinit** [comment]
Finally had time to fix this.

**Review by RossComputerGuy — CHANGES_REQUESTED** [review]
Probably should put this in a let block.

File: pkgs/applications/science/logic/klee/default.nix [hunk 5]
```diff
@@ -14,6 +12,27 @@
 , gtest
 , lit
 
+# The chosen version of LLVM.
+, chosenLLVMPackages ? let
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:16]: Probably should put this in a let block.

**Review by RossComputerGuy — CHANGES_REQUESTED** [review]

File: pkgs/applications/science/logic/klee/default.nix (hunk 5 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix]: Just defaulted it to null so they can still override it if they really want to, and moved this and the below to the let block.

**Review by RossComputerGuy — APPROVED** [review]

**Comment by RossComputerGuy** [comment]
Result of `nixpkgs-review pr 304204` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by numinit** [comment]
Rebased; should be good to go

File: pkgs/applications/science/logic/klee/default.nix [hunk 6]
```diff
@@ -30,19 +29,43 @@
...
+# Klee uclibc. Defaults to the bundled version.
+, kleeuClibc ? null
+
+# Extra klee-uclibc config for the default klee-uclibc.
 , extraKleeuClibcConfig ? {}
 }:
 
 let
+  # The chosen version of LLVM.
+  chosenLLVMPackages = let
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:41]: Why not use an `assert` instead of making an entirely new variable?

File: pkgs/applications/science/logic/klee/default.nix [hunk 7]
```diff
@@ -30,19 +29,43 @@
...
+    maxLLVMVersion = 16;
+  in
+  if versionAtLeast version  (toString minLLVMVersion)
+     && versionOlder version (toString (maxLLVMVersion + 1)) then
+    llvmPackages
+  else
+    throw "Unsupported LLVM version: ${version} (must be in ${toString minLLVMVersion} to ${toString maxLLVMVersion})";
+
+  # The chosen version of klee-uclibc.
+  chosenKleeuClibc =
```
Comment (RossComputerGuy) [inline pkgs/applications/science/logic/klee/default.nix:56]: Why not use an `assert` instead of making an entirely new variable?

**Review by RossComputerGuy — DISMISSED** [review]

File: pkgs/applications/science/logic/klee/default.nix (hunk 6 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix:41]: Mainly because I wanted to throw a useful error.

File: pkgs/applications/science/logic/klee/default.nix (hunk 7 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix:56]: This actually is something else: it's defaulting it to the builtin klee-uclibc if the callPackage input is null.

Comment (RossComputerGuy) [reply pkgs/applications/science/logic/klee/default.nix:56]: Oh, didn't quite see that.

File: pkgs/applications/science/logic/klee/default.nix (hunk 6 above)
Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix]: Fixed.

Comment (numinit) [reply pkgs/applications/science/logic/klee/default.nix]: `error: assertion '((versionAtLeast version "11") && (versionOlder version "17"))' failed` - actually a nice error :-)

**Review by RossComputerGuy — APPROVED** [review]
